File: rules-cli/src/main/java/RuleEvaluator.java
import io.github.iamrenny.ruleflow.visitors.Visitor;
import io.github.iamrenny.ruleflow.RuleFlowLanguageLexer;
import io.github.iamrenny.ruleflow.RuleFlowLanguageParser;
import org.antlr.v4.runtime.BaseErrorListener;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.ParserRuleContext;
import org.antlr.v4.runtime.RecognitionException;
import org.antlr.v4.runtime.Recognizer;
import org.antlr.v4.runtime.tree.ParseTree;

import java.util.Map;
import java.util.Set;

public class RuleEvaluator {
    private final String inputRule;
    private final Map<String, ?> data;
    private final Map<String, Set<String>> lists;
    private final SyntaxErrorListener syntaxErrorListener = new SyntaxErrorListener();

    public RuleEvaluator(String inputRule, Map<String, ?> data, Map<String, Set<String>> lists) {
        this.inputRule = inputRule;
        this.data = data;
        this.lists = lists;
    }

    public String evaluate() {
        var input = CharStreams.fromString(inputRule);
        var lexer = new RuleFlowLanguageLexer(input);
        var tokens = new CommonTokenStream(lexer);
        var parser = new RuleFlowLanguageParser(tokens);

        // Remove default error listeners and add custom listener
        parser.removeErrorListeners();
        parser.addErrorListener(syntaxErrorListener);

        ParseTree ctx = parser.expr();
        ParserRuleContext parserContext = (ParserRuleContext) ctx;

        // If there is a syntax error, return the error message
        if (syntaxErrorListener.isFailed()) {
            return "error: " + syntaxErrorListener.getErrorMessage() +
                    "\n" + tokens.getText(parserContext.getStart(), parserContext.getStop());
        }

        var ruleEvaluator = new Visitor(data, lists, data);

        try {
            return ruleEvaluator.visit(ctx).toString();
        } catch (Exception ex) {
            System.out.println("Error while evaluating rule " + tokens.getText(parserContext.getStart(), parserContext.getStop()) + ": " + ex.getMessage());
            return "error: " + ex.getMessage() + "\n" + tokens.getText(parserContext.getStart(), parserContext.getStop());
        }
    }
}

class SyntaxErrorListener extends BaseErrorListener {
    private boolean failed = false;
    private String errorMessage;

    public boolean isFailed() {
        return failed;
    }

    public String getErrorMessage() {
        return errorMessage;
    }

    @Override
    public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine, String msg, RecognitionException e) {
        errorMessage = "Error at line " + line + ":" + charPositionInLine + " - " + msg;
        failed = true;
    }
}

File: rules-cli/src/main/kotlin/RuleEvaluator.kt
import com.rappi.analang.visitors.Visitor
import com.rappi.fraud.analang.ANALexer
import com.rappi.fraud.analang.ANAParser
import org.antlr.v4.runtime.BaseErrorListener
import org.antlr.v4.runtime.CharStreams
import org.antlr.v4.runtime.CommonTokenStream
import org.antlr.v4.runtime.RecognitionException
import org.antlr.v4.runtime.Recognizer
import org.slf4j.LoggerFactory

class RuleEvaluator(val inputRule: String, private val data: Map<String, *>, private val lists:  Map<String, Set<String>>) {
    val syntaxErrorListener = SyntaxErrorListener()

    fun evaluate(): String {
        val input = CharStreams.fromString(inputRule)
        val lexer = ANALexer(input)
        val tokens = CommonTokenStream(lexer)
        val parser = ANAParser(tokens)
        parser.removeErrorListeners()  // Remove the default error listeners
        parser.addErrorListener(syntaxErrorListener)
        val ctx = parser.expr()

        if(syntaxErrorListener.failed)
            return "error: ${syntaxErrorListener.errorMessage!!}" +
                    "\n${tokens.getText(ctx.start, ctx.stop)}"


        val ruleEvaluator = Visitor(data, lists, data)

        return try {
            return ruleEvaluator.visit(ctx).toString()
        }  catch (ex: Exception) {
            println(
                "Error while evaluating rule ${tokens.getText(ctx.start, ctx.stop)}: ${ex.message}"
            )
            "error: ${ex.message!!}\n${tokens.getText(ctx.start, ctx.stop)}"
        }
    }
}
class SyntaxErrorListener : BaseErrorListener() {
    var failed: Boolean = false
    var errorMessage: String? = null
    override fun syntaxError(recognizer: Recognizer<*, *>?,
                             offendingSymbol: Any?,
                             line: Int,
                             charPositionInLine: Int,
                             msg: String?,
                             e: RecognitionException?) {
        errorMessage = "Error at line $line:$charPositionInLine - $msg"
        failed = true
    }
}
